pkg/apis/crd/v1alpha2/types.go

```diff
@@ -389,3 +389,110 @@ type IPPoolList struct {
 
 	Items []IPPool `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// TrafficControl allows mirroring or redirecting the traffic Pods send or receive. It enables users to monitor and
+// analyze Pod traffic, and to enforce custom network protections for Pods with fine-grained control over network
+// traffic.
+type TrafficControl struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard metadata of the object.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of TrafficControl.
+	Spec TrafficControlSpec `json:"spec"`
+}
+
+type TrafficControlSpec struct {
+	// AppliedTo selects Pods to which the traffic control configuration will be applied.
+	AppliedTo AppliedTo `json:"appliedTo"`
+
+	// The direction of traffic that should be matched. It can be Ingress, Egress, or Both.
+	Direction Direction `json:"direction"`
+
+	// The action that should be taken for the traffic. It can be Redirect or Mirror.
+	Action TrafficControlAction `json:"action"`
+
+	// The device to which the traffic should be redirected or mirrored.
+	TargetPort OVSPort `json:"targetPort"`
+
+	// The device from which the the traffic will be sent back to OVS. It should only be set for Redirect action.
+	ReturnPort *OVSPort `json:"returnPort,omitempty"`
+}
+
+type Direction string
+
+const (
+	DirectionIngress Direction = "Ingress"
+	DirectionEgress  Direction = "Egress"
+	DirectionBoth    Direction = "Both"
+)
+
+type TrafficControlAction string
+
+const (
+	ActionRedirect TrafficControlAction = "Redirect"
+	ActionMirror   TrafficControlAction = "Mirror"
+)
+
+type OVSPortType string
+
+const (
+	OVSPortTypeInternal OVSPortType = "Internal"
+	OVSPortTypeDevice   OVSPortType = "Device"
+	OVSPortTypeVXLAN    OVSPortType = "VXLAN"
+	OVSPortTypeGENEVE   OVSPortType = "GENEVE"
+	OVSPortTypeGRE      OVSPortType = "GRE"
+	OVSPortTypeERSPAN   OVSPortType = "ERSPAN"
+)
+
+// OVSPort represents an OVS port that can be used as the target of traffic mirroring or redirecting, and the return
+// port of traffic redirecting. Multiple TrafficControl objects can use same port, however, the ports with same name
```

Comment:
Do we plan to support: "Multiple TrafficControl objects can use same port" in the current version?

Reply:
I think there is no harm to support it with the statement that ports with same name must have same configuration:
If we don't allow it, it requires more code to validate the port is not duplicate with any of existing ports, or how we say it's not allowed?
If we allow it with the precondition, port sharing can work if it's configured correctly.

Reply:
We need to validate the configuration is the same, and we need to track the reference to know when to delete the port.

Reply:
Ok, we had long discussion on this and I think we are on the same page. How about you choose one based on complexity and your preference. If we want to remove port name for tunnel, then we should go your way of:

```
type TrafficControlPort struct {
	OVSInternal *OVSInternalPort `json:"ovsInternal,omitempty"`
	Device *DevicePort `json:"device,omitempty"`
	VXLAN *TunnelPort `json:"vxlan,omitempty"`
	GENEVE *TunnelPort `json:"geneve,omitempty"`
	GRE *TunnelPort `json:"gre,omitempty"`
	ERSPAN *ERSPANPort `json:"erspan,omitempty"`
}

// Maybe we just need a string named "InternalPort" in TrafficControlPort?
type InternalPort struct {
        // The name of the OVS internal port.
	Name string `json:"name"`
}

// How about NetworkDevice?
type DevicePort struct {
	Name string `json:"name"`
}

// Probably do not call it TunnelPort then, but just Tunnel?
type TunnelPort struct {
	RemoteIP string `json:"remoteIP"`
	TunnelID int64 `json:"tunnelID,omitempty"`
}
```

Again, my major problem is then not straightforward to support pre-created port, but maybe we can add a filed like PrecreatedPort/ExistingPort, etc. for that when we need to support it.